doc/cacaview.1
.TH cacaview 1 "2003-11-30" "libcaca"
.SH NAME
cacaview \- ASCII image browser
.SH SYNOPSIS
.B cacaview [FILE...]
.RI
.SH DESCRIPTION
.B cacaview
is a lightweight text mode image viewer. It renders images using colour
ASCII characters. It is a powerful add-on to famous console programs such
as the mutt email client, the slrn newsreader and the links or w3m web
browsers.
.PP
.B cacaview
can load the most widespread image formats: PNG, JPEG, GIF, PNG, BMP etc.
.PP
You can zoom and scroll the image around for more details, and choose four
different dithering modes. All commands are accessible through a single
key press.
.SH KEYS
.TP
.B ?
show the help screen
.TP
.B n, p
switch to next image, previous image
.TP
.B Left, Right, Up, Down or h, l, k, j
scroll the image around
.TP
.B +, -
zoom in and out
.TP
.B z
reset the zoom level to normal
.TP
.B f
switch fullscreen mode (hide/show menu and status bars)
.TP
.B d
toggle the dithering mode (no dithering, 4x4 ordered dithering, 8x8 ordered
dithering and random dithering)
.TP
.B q
exit the program
.SH EXAMPLE
cacaview /usr/share/pixmaps/*.*
.SH BUGS
There is no support for aspect ratio yet. Also, since there is no way
yet to load an image from
.B cacaview
it is completely useless when run without an argument.
.SH SEE ALSO
img2txt(1)
.SH AUTHOR
This manual page was written by Sam Hocevar <sam@hocevar.net>.


NEWS
/* $Id$ */ /** \page libcaca-news Libcaca news

\section news0_99_beta17 Changes between 0.9.beta17 and 0.99.beta16

 - new dirty rectangle framework for accelerated rendering
 - new <conio.h> compatible header for old DOS software porting
 - XCode and Visual Studio build improvements
 - improved HTML output
 - improved Ruby and .NET bindings
 - new comprehensive PHP bindings
 - Cocoa driver enhancements
 - documentation updates
 - minor improvements and bugfixes

\section news0_99_beta16 Changes between 0.9.beta16 and 0.99.beta15

 - many build fixes, especially for nonstandard platforms

\section news0_99_beta15 Changes between 0.9.beta15 and 0.99.beta14

 - libcucul was merged back into libcaca for more clarity

\section news0_99_beta14 Changes between 0.9.beta14 and 0.99.beta13

 - internal FIGlet font support
 - use C99 types in public headers
 - runtime support for output drivers
 - BBcode export support

\section news0_99_beta13 Changes between 0.9.beta13 and 0.99.beta12

 - device-dependent cursor support
 - event API rewrite
 - minor API improvements and extensions
 - img2txt improvements
 - Ruby bindings
 - Massive C# bindings improvements
 - Python sample code
 - Visual Studio build solution

\section news0_99_beta12 Changes between 0.9.beta12 and 0.99.beta11

 - support for 90-degree canvas rotation
 - better behaviour when trying to output Unicode on an ASCII terminal
 - the built-in font now supports the Geometric Shapes, Halfwidth and Fullwidth
   Forms and Miscellaneous Symbols Unicode blocks
 - new rotozoom effect in cacademo
 - Cocoa output driver for Mac OS X
 - preliminary .NET bindings
 - many bugfixes and documentation changes

\section news0_99_beta11 Changes between 0.9.beta11 and 0.99.beta10

 - fixed compilation of the C++ bindings
 - fixed bugs in cucul_import_memory(), cucul_set_canvas_size()
 - implemented caca_set_display_title() for ncurses and S-Lang
 - minor bugfixes

\section news0_99_beta10 Changes between 0.9.beta10 and 0.99.beta9

 - new debug mode
 - blitting canvases now makes use of the canvas' handle coordinates
 - import functions can read streamed data
 - attribute to colorspace transformations
 - added katakana and hiragana glyphs to the built-in font
 - many bugfixes and documentation changes

\section news0_99_beta9 Changes between 0.9.beta9 and 0.99.beta8

 - support for blink, bold, italics and underline attributes
 - allow to import and export zero-sized canvases
 - fixed Imlib2 support in cacaview
 - fixed buffer overlows in the file importer
 - big documentation updates

\section news0_99_beta8 Changes between 0.9.beta8 and 0.99.beta7

 - allow to build the X11 and GL drivers as separate plugins
 - support for fullwidth Unicode characters
 - improved cucul_flip() and cucul_rotate()
 - minor bugfixes and documentation updates

\section news0_99_beta7 Changes between 0.9.beta7 and 0.99.beta6

 - transparency support in the UTF-8 importer and exporter
 - optimised triangle fill routine
 - updated C++ bindings

\section news0_99_beta6 Changes between 0.9.beta6 and 0.99.beta5

 - ANSI importer now handles transparency and UTF-8
 - Unicode support was broken for about 10% of the set
 - various memory leaks were fixed

\section news0_99_beta5 Changes between 0.9.beta5 and 0.99.beta4

 - implemented cucul_getchar() and cucul_get_color()
 - handle transparency in the IRC export
 - new cropping and expanding filters
 - full Unicode support in the OpenGL driver
 - portability fixes for 64-bit architectures, Win32 and MS-DOS
 - all demos except cacafire were merged into cacademo

\section news0_99_beta4 Changes between 0.9.beta4 and 0.99.beta3

 - added a compatibility layer for pre-1.x libcaca applications
 - fixed manpage generation
 - minor bugfixes and documentation updates

\section news0_99_beta3 Changes between 0.9.beta3 and 0.99.beta2

 - libcaca functions use errno for error reporting
 - updated C++ bindings
 - minor improvements, bugfixes and documentation updates

\section news0_99_beta2 Changes between 0.9.beta2 and 0.99.beta1

 - ANSI importer
 - functions use errno for error reporting
 - updated C++ bindings
 - .NET bindings
 - cacadraw, an ANSI viewer that will evolve into an editor
 - Unicode input and output support for SLang and ncurses
 - built-in fonts work on Win32

\section news0_99_beta1 Changes between 0.9 and 0.99.beta1

 - license switched to WTFPL
 - libcaca was split into libcucul, a standalone text manipulation backend,
    and libcaca, the display and user input frontend
 - Unicode support
 - TrueColor (more than 16 colours) support
 - Floyd-Steinberg dithering
 - gamma correction
 - export functions for HTML, IRC, ANSI, SVG, PostScript, TGA...
 - builtin fonts for device-independent bitmap output
 - various text transformation routines (rotation, horizontal flip...)
 - OpenGL renderer
 - kernel mode to build libcaca programs into a bootable x86 kernel
 - cacaserver, a telnet server that can be hooked to libcaca applications
 - img2irc, an image to IRC conversion utility

\section news0_9 Changes between 0.8 and 0.9

 - fix for a buffer overflow in the line rendering
 - fixed resizing in the ncurses and slang drivers
 - aspect ratio and finer zoom support in cacaview
 - minor compilation fixes

\section news0_8 Changes between 0.7 and 0.8

 - window resizing support
 - native Win32 port
 - autorepeat emulation in the ncurses and slang drivers
 - support for more keycodes in the ncurses and slang drivers
 - cacaplas, a plasma animation example
 - cacamoir, a moire circles animation example
 - MSVC project file

\section news0_7 Changes between 0.6 and 0.7

 - many bugfixes in the event handling
 - cacaball, a metaball animation example

\section news0_6 Changes between 0.5 and 0.6

 - 30% speed increase in the bitmap rendering routine
 - mouse support and various speed optimisations in the X11 driver
 - X11 is now the preferred driver
 - improved documentation
 - minor bugfixes


src/cacaview.c
/*
 *  cacaview      image viewer for libcaca
 *  Copyright (c) 2003-2006 Sam Hocevar <sam@hocevar.net>
 *                All Rights Reserved
 *
 *  $Id$
 *
 *  This program is free software. It comes without any warranty, to
 *  the extent permitted by applicable law. You can redistribute it
 *  and/or modify it under the terms of the Do What The Fuck You Want
 *  To Public License, Version 2, as published by Sam Hocevar. See
 *  http://sam.zoy.org/wtfpl/COPYING for more details.
 */

#include "config.h"

#if !defined(__KERNEL__)
#   include <stdio.h>
#   include <string.h>
#   include <stdlib.h>
#endif

#if defined(HAVE_SLEEP)
#   include <windows.h>
#endif

#include "caca.h"

#include "common-image.h"

/* Local macros */
#define MODE_IMAGE 1
#define MODE_FILES 2

#define STATUS_DITHERING 1
#define STATUS_ANTIALIASING 2
#define STATUS_BACKGROUND 3

#define ZOOM_FACTOR 1.08f
#define ZOOM_MAX 50
#define GAMMA_FACTOR 1.04f
#define GAMMA_MAX 100
#define GAMMA(g) (((g) < 0) ? 1.0 / gammatab[-(g)] : gammatab[(g)])
#define PAD_STEP 0.15

/* libcaca/libcaca contexts */
caca_canvas_t *cv; caca_display_t *dp;

/* Local functions */
static void print_status(void);
static void print_help(int, int);
static void set_zoom(int);
static void set_gamma(int);
static void draw_checkers(int, int, int, int);

/* Local variables */
struct image *im = NULL;

float zoomtab[ZOOM_MAX + 1];
float gammatab[GAMMA_MAX + 1];
float xfactor = 1.0, yfactor = 1.0, dx = 0.5, dy = 0.5;
int zoom = 0, g = 0, fullscreen = 0, mode, ww, wh;

int main(int argc, char **argv)
{
    char const * const * algos = caca_get_dither_algorithm_list(NULL);
    int dither_algorithm = 0;

    int quit = 0, update = 1, help = 0, status = 0;
    int reload = 0;

    char **list = NULL;
    int current = 0, items = 0, opts = 1;
    int i;

    /* Initialise libcaca */
    cv = caca_create_canvas(0, 0);
    if(!cv)
    {
        fprintf(stderr, "%s: unable to initialise libcaca\n", argv[0]);
        return 1;
    }

    dp = caca_create_display(cv);
    if(!dp)
    {
        fprintf(stderr, "%s: unable to initialise libcaca\n", argv[0]);
        return 1;
    }

    /* Set the window title */
    caca_set_display_title(dp, "cacaview");

    ww = caca_get_canvas_width(cv);
    wh = caca_get_canvas_height(cv);

    /* Fill the zoom table */
    zoomtab[0] = 1.0;
    for(i = 0; i < ZOOM_MAX; i++)
        zoomtab[i + 1] = zoomtab[i] * ZOOM_FACTOR;

    /* Fill the gamma table */
    gammatab[0] = 1.0;
    for(i = 0; i < GAMMA_MAX; i++)
        gammatab[i + 1] = gammatab[i] * GAMMA_FACTOR;

    /* Load items into playlist */
    for(i = 1; i < argc; i++)
    {
        /* Skip options except after `--' */
        if(opts && argv[i][0] == '-')
        {
            if(argv[i][1] == '-' && argv[i][2] == '\0')
                opts = 0;
            continue;
        }

        /* Add argv[i] to the list */
        if(items)
            list = realloc(list, (items + 1) * sizeof(char *));
        else
            list = malloc(sizeof(char *));
        list[items] = argv[i];
        items++;

        reload = 1;
    }

    /* Go ! */
    while(!quit)
    {
        caca_event_t ev;
        unsigned int const event_mask = CACA_EVENT_KEY_PRESS
                                      | CACA_EVENT_RESIZE
                                      | CACA_EVENT_MOUSE_PRESS
                                      | CACA_EVENT_QUIT;
        unsigned int new_status = 0, new_help = 0;
        int event;

        if(update)
            event = caca_get_event(dp, event_mask, &ev, 0);
        else
            event = caca_get_event(dp, event_mask, &ev, -1);

        while(event)
        {
            if(caca_get_event_type(&ev) & CACA_EVENT_MOUSE_PRESS)
            {
                if(caca_get_event_mouse_button(&ev) == 1)
                {
                    if(items) current = (current + 1) % items;
                    reload = 1;
                }
                if(caca_get_event_mouse_button(&ev) == 2)
                {
                    if(items) current = (items + current - 1) % items;
                    reload = 1;
                }
            }
            else if(caca_get_event_type(&ev) & CACA_EVENT_KEY_PRESS)
                switch(caca_get_event_key_ch(&ev))
            {
            case 'n':
            case 'N':
                if(items) current = (current + 1) % items;
                reload = 1;
                break;
            case 'p':
            case 'P':
                if(items) current = (items + current - 1) % items;
                reload = 1;
                break;
            case 'f':
            case 'F':
            case CACA_KEY_F11:
                fullscreen = ~fullscreen;
                update = 1;
                set_zoom(zoom);
                break;
#if 0 /* FIXME */
            case 'b':
                i = 1 + caca_get_feature(cv, CACA_BACKGROUND);
                if(i > CACA_BACKGROUND_MAX) i = CACA_BACKGROUND_MIN;
                caca_set_feature(cv, i);
                new_status = STATUS_BACKGROUND;
                update = 1;
                break;
            case 'B':
                i = -1 + caca_get_feature(cv, CACA_BACKGROUND);
                if(i < CACA_BACKGROUND_MIN) i = CACA_BACKGROUND_MAX;
                caca_set_feature(cv, i);
                new_status = STATUS_BACKGROUND;
                update = 1;
                break;
            case 'a':
                i = 1 + caca_get_feature(cv, CACA_ANTIALIASING);
                if(i > CACA_ANTIALIASING_MAX) i = CACA_ANTIALIASING_MIN;
                caca_set_feature(cv, i);
                new_status = STATUS_ANTIALIASING;
                update = 1;
                break;
            case 'A':
                i = -1 + caca_get_feature(cv, CACA_ANTIALIASING);
                if(i < CACA_ANTIALIASING_MIN) i = CACA_ANTIALIASING_MAX;
                caca_set_feature(cv, i);
                new_status = STATUS_ANTIALIASING;
                update = 1;
                break;
#endif
            case 'd':
                dither_algorithm++;
                if(algos[dither_algorithm * 2] == NULL)
                    dither_algorithm = 0;
                caca_set_dither_algorithm(im->dither,
                                           algos[dither_algorithm * 2]);
                new_status = STATUS_DITHERING;
                update = 1;
                break;
            case 'D':
                dither_algorithm--;
                if(dither_algorithm < 0)
                    while(algos[dither_algorithm * 2 + 2] != NULL)
                        dither_algorithm++;
                caca_set_dither_algorithm(im->dither,
                                           algos[dither_algorithm * 2]);
                new_status = STATUS_DITHERING;
                update = 1;
                break;
            case '+':
                update = 1;
                set_zoom(zoom + 1);
                break;
            case '-':
                update = 1;
                set_zoom(zoom - 1);
                break;
            case 'G':
                update = 1;
                set_gamma(g + 1);
                break;
            case 'g':
                update = 1;
                set_gamma(g - 1);
                break;
            case 'x':
            case 'X':
                update = 1;
                set_zoom(0);
                set_gamma(0);
                break;
            case 'k':
            case 'K':
            case CACA_KEY_UP:
                if(yfactor > 1.0) dy -= PAD_STEP / yfactor;
                if(dy < 0.0) dy = 0.0;
                update = 1;
                break;
            case 'j':
            case 'J':
            case CACA_KEY_DOWN:
                if(yfactor > 1.0) dy += PAD_STEP / yfactor;
                if(dy > 1.0) dy = 1.0;
                update = 1;
                break;
            case 'h':
            case 'H':
            case CACA_KEY_LEFT:
                if(xfactor > 1.0) dx -= PAD_STEP / xfactor;
                if(dx < 0.0) dx = 0.0;
                update = 1;
                break;
            case 'l':
            case 'L':
            case CACA_KEY_RIGHT:
                if(xfactor > 1.0) dx += PAD_STEP / xfactor;
                if(dx > 1.0) dx = 1.0;
                update = 1;
                break;
            case '?':
                new_help = !help;
                update = 1;
                break;
            case 'q':
            case 'Q':
            case CACA_KEY_ESCAPE:
                quit = 1;
                break;
            }
            else if(caca_get_event_type(&ev) == CACA_EVENT_RESIZE)
            {
                caca_refresh_display(dp);
                ww = caca_get_event_resize_width(&ev);
                wh = caca_get_event_resize_height(&ev);
                update = 1;
                set_zoom(zoom);
            }
            else if(caca_get_event_type(&ev) & CACA_EVENT_QUIT)
                quit = 1;

            if(status || new_status)
                status = new_status;

            if(help || new_help)
                help = new_help;

            event = caca_get_event(dp, CACA_EVENT_KEY_PRESS, &ev, 0);
        }

        if(items && reload)
        {
            char *buffer;
            int len = strlen(" Loading `%s'... ") + strlen(list[current]);

            if(len < ww + 1)
                len = ww + 1;

            buffer = malloc(len);

            sprintf(buffer, " Loading `%s'... ", list[current]);
            buffer[ww] = '\0';
            caca_set_color_ansi(cv, CACA_WHITE, CACA_BLUE);
            caca_put_str(cv, (ww - strlen(buffer)) / 2, wh / 2, buffer);
            caca_refresh_display(dp);
            ww = caca_get_canvas_width(cv);
            wh = caca_get_canvas_height(cv);

            if(im)
                unload_image(im);
            im = load_image(list[current]);
            reload = 0;

            /* Reset image-specific runtime variables */
            dx = dy = 0.5;
            update = 1;
            set_zoom(0);
            set_gamma(0);

            free(buffer);
        }

        caca_set_color_ansi(cv, CACA_WHITE, CACA_BLACK);
        caca_clear_canvas(cv);

        if(!items)
        {
            caca_set_color_ansi(cv, CACA_WHITE, CACA_BLUE);
            caca_printf(cv, ww / 2 - 5, wh / 2, " No image. ");
        }
        else if(!im)
        {
#if defined(USE_IMLIB2)
#   define ERROR_STRING " Error loading `%s'. "
#else
#   define ERROR_STRING " Error loading `%s'. Only BMP is supported. "
#endif
            char *buffer;
            int len = strlen(ERROR_STRING) + strlen(list[current]);

            if(len < ww + 1)
                len = ww + 1;

            buffer = malloc(len);

            sprintf(buffer, ERROR_STRING, list[current]);
            buffer[ww] = '\0';
            caca_set_color_ansi(cv, CACA_WHITE, CACA_BLUE);
            caca_put_str(cv, (ww - strlen(buffer)) / 2, wh / 2, buffer);
            free(buffer);
        }
        else
        {
            float xdelta, ydelta;
            int y, height;

            y = fullscreen ? 0 : 1;
            height = fullscreen ? wh : wh - 3;

            xdelta = (xfactor > 1.0) ? dx : 0.5;
            ydelta = (yfactor > 1.0) ? dy : 0.5;

            draw_checkers(ww * (1.0 - xfactor) / 2,
                          y + height * (1.0 - yfactor) / 2,
                          ww * xfactor, height * yfactor);

            caca_dither_bitmap(cv, ww * (1.0 - xfactor) * xdelta,
                            y + height * (1.0 - yfactor) * ydelta,
                            ww * xfactor + 1, height * yfactor + 1,
                            im->dither, im->pixels);
        }

        if(!fullscreen)
        {
            print_status();

            caca_set_color_ansi(cv, CACA_LIGHTGRAY, CACA_BLACK);
            switch(status)
            {
                case STATUS_DITHERING:
                    caca_printf(cv, 0, wh - 1, "Dithering: %s",
                                 caca_get_dither_algorithm(im->dither));
                    break;
#if 0 /* FIXME */
                case STATUS_ANTIALIASING:
                    caca_printf(cv, 0, wh - 1, "Antialiasing: %s",
                  caca_get_feature_name(caca_get_feature(cv, CACA_ANTIALIASING)));
                    break;
                case STATUS_BACKGROUND:
                    caca_printf(cv, 0, wh - 1, "Background: %s",
                  caca_get_feature_name(caca_get_feature(cv, CACA_BACKGROUND)));
                    break;
#endif
            }
        }

        if(help)
        {
            print_help(ww - 26, 2);
        }

        caca_refresh_display(dp);
        update = 0;
    }

    /* Clean up */
    if(im)
        unload_image(im);
    caca_free_display(dp);
    caca_free_canvas(cv);

    return 0;
}

static void print_status(void)
{
    caca_set_color_ansi(cv, CACA_WHITE, CACA_BLUE);
    caca_draw_line(cv, 0, 0, ww - 1, 0, ' ');
    caca_draw_line(cv, 0, wh - 2, ww - 1, wh - 2, '-');
    caca_put_str(cv, 0, 0, "q:Quit  np:Next/Prev  +-x:Zoom  gG:Gamma  "
                            "hjkl:Move  d:Dither  a:Antialias");
    caca_put_str(cv, ww - strlen("?:Help"), 0, "?:Help");
    caca_printf(cv, 3, wh - 2, "cacaview %s", PACKAGE_VERSION);
    caca_printf(cv, ww - 30, wh - 2, "(gamma: %#.3g)", GAMMA(g));
    caca_printf(cv, ww - 14, wh - 2, "(zoom: %s%i)", zoom > 0 ? "+" : "", zoom);

    caca_set_color_ansi(cv, CACA_LIGHTGRAY, CACA_BLACK);
    caca_draw_line(cv, 0, wh - 1, ww - 1, wh - 1, ' ');
}

static void print_help(int x, int y)
{
    static char const *help[] =
    {
        " +: zoom in              ",
        " -: zoom out             ",
        " g: decrease gamma       ",
        " G: increase gamma       ",
        " x: reset zoom and gamma ",
        " ----------------------- ",
        " hjkl: move view         ",
        " arrows: move view       ",
        " ----------------------- ",
        " a: antialiasing method  ",
        " d: dithering method     ",
        " b: background mode      ",
        " ----------------------- ",
        " ?: help                 ",
        " q: quit                 ",
        NULL
    };

    int i;

    caca_set_color_ansi(cv, CACA_WHITE, CACA_BLUE);

    for(i = 0; help[i]; i++)
        caca_put_str(cv, x, y + i, help[i]);
}

static void set_zoom(int new_zoom)
{
    int height;

    if(!im)
        return;

    zoom = new_zoom;

    if(zoom > ZOOM_MAX) zoom = ZOOM_MAX;
    if(zoom < -ZOOM_MAX) zoom = -ZOOM_MAX;

    ww = caca_get_canvas_width(cv);
    height = fullscreen ? wh : wh - 3;

    xfactor = (zoom < 0) ? 1.0 / zoomtab[-zoom] : zoomtab[zoom];
    yfactor = xfactor * ww / height * im->h / im->w
               * caca_get_canvas_height(cv) / caca_get_canvas_width(cv)
               * caca_get_display_width(dp) / caca_get_display_height(dp);

    if(yfactor > xfactor)
    {
        float tmp = xfactor;
        xfactor = tmp * tmp / yfactor;
        yfactor = tmp;
    }
}

static void set_gamma(int new_gamma)
{
    if(!im)
        return;

    g = new_gamma;

    if(g > GAMMA_MAX) g = GAMMA_MAX;
    if(g < -GAMMA_MAX) g = -GAMMA_MAX;

    caca_set_dither_gamma(im->dither,
                           (g < 0) ? 1.0 / gammatab[-g] : gammatab[g]);
}

static void draw_checkers(int x, int y, int w, int h)
{
    int xn, yn;

    if(x + w > (int)caca_get_canvas_width(cv))
        w = caca_get_canvas_width(cv) - x;
    if(y + h > (int)caca_get_canvas_height(cv))
        h = caca_get_canvas_height(cv) - y;

    for(yn = y > 0 ? y : 0; yn < y + h; yn++)
        for(xn = x > 0 ? x : 0; xn < x + w; xn++)
    {
        if((((xn - x) / 5) ^ ((yn - y) / 3)) & 1)
            caca_set_color_ansi(cv, CACA_LIGHTGRAY, CACA_DARKGRAY);
        else
            caca_set_color_ansi(cv, CACA_DARKGRAY, CACA_LIGHTGRAY);
        caca_put_char(cv, xn, yn, ' ');
    }
}

